src/System.Diagnostics.Tracing.Telemetry/src/System/Diagnostics/Tracing/TelemetryListener.cs

```diff
@@ -0,0 +1,249 @@
+        /// </summary>
+        /// <returns></returns>
+        public override string ToString()
+        {
+            return Name;
+        }
+
+        #region private
+
+        // NotificationSource implementation
+        public override bool IsEnabled(string telemetryName)
+        {
+            for (var curSubscription = m_subscriptions; curSubscription != null; curSubscription = curSubscription.Next)
+            {
+                if (curSubscription.IsEnabled == null || curSubscription.IsEnabled(telemetryName))
+                    return true;
+            }
+            return false;
+        }
+
+        public override void WriteTelemetry(string telemetryName, object parameters)
+        {
+            for (var curSubscription = m_subscriptions; curSubscription != null; curSubscription = curSubscription.Next)
+                curSubscription.Observer.OnNext(new KeyValuePair<string, object>(telemetryName, parameters));
+        }
+
+        // Note that Subscriptions are READ ONLY.   This means you never update any fields (even on removal!)
+        private class Subscription : IDisposable
+        {
+            internal IObserver<KeyValuePair<string, object>> Observer;
+            internal Predicate<string> IsEnabled;
+            internal TelemetryListener Owner;          // The TelemetryListener this is a subscription for.  
+            internal Subscription Next;                // Linked list of subscribers
+
+            public void Dispose()
+            {
+                // TO keep this lock free and easy to analyze, the linked list is READ ONLY.   Thus we copy
+                for (;;)
+                {
+                    Subscription subscriptions = Owner.m_subscriptions;
+                    Subscription newSubscriptions = Remove(subscriptions, this);    // Make a new list, with myself removed.  
+
+                    // try to update, but if someone beat us to it, then retry.  
+                    if (Interlocked.CompareExchange(ref Owner.m_subscriptions, newSubscriptions, subscriptions) == subscriptions)
+                    {
+#if DEBUG
+                        var cur = subscriptions;
+                        while (cur != null)
+                        {
+                            if (cur.Observer == this)
+                                throw new Exception("Did not remove subscription!");
+                            cur = cur.Next;
+                        }
+#endif
+                        break;
+                    }
+                }
+            }
+
+            // Create a new linked list where 'subscription has been removed from the linked list of 'subscriptions'. 
+            private static Subscription Remove(Subscription subscriptions, Subscription subscription)
+            {
+                if (subscriptions == null)
+                {
+#if DEBUG
+                    throw new Exception("Could not find subscription");
```

Comment:
If taking a dependency on System.Diagnostics.Debug it not a problem then yes, that is the way to do it.